### PR TITLE
Add basics API reference

### PR DESCRIPTION
**Pull request by SimonHeybrock**
This is not 100% complete as we are still working on a more user-friendly solution in some cases, but these are the most important bits.

### CODE REVIEW

File: docs/api-reference/index.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+## Submodules
+
+```{eval-rst}
+.. autosummary::
+   :toctree: ../generated/modules
+   :template: module-template.rst
+   :recursive:
+
+   beam_center_finder
```
Comment (nvaytet) [inline docs/api-reference/index.md:32]: Is there a way we can improve the long function names? Maybe only have the function name and not the submodule path? But I guess that path is important information to have...
Screenshot at 2023-10-04 15-16-28

File: docs/api-reference/index.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+   sans2d
+   types
```
Comment (nvaytet) [inline docs/api-reference/index.md:34]: Was it intentional that some entries under "Module attributes" and others under "Classes"? I don't really seem to get the logic separating the two sections.

File: docs/api-reference/index.md (hunk 1 above)
Comment (SimonHeybrock) [reply docs/api-reference/index.md:32]: I suppose we can shorten, `beam_center_finder` -> `beam_center`, or remove the duplicate `beam_center_` prefix from the 2 finders?

File: docs/api-reference/index.md (hunk 2 above)
Comment (SimonHeybrock) [reply docs/api-reference/index.md:34]: It is not intentional, but this is what we get from the generic template, and due to the fact that some domain types are using `NewType` while others are not (are actual classes). I don't know how to fix that. Maybe the template could be rewritten?

File: docs/api-reference/index.md (hunk 1 above)
Comment (nvaytet) [reply docs/api-reference/index.md:32]: > remove the duplicate beam_center_ prefix from the 2 finders?

would work in this case. But I guess it won't save us from future functions with long names.

Comment (SimonHeybrock) [reply docs/api-reference/index.md:32]: Maybe having shorter module names is good style generally. Should we just rename `beam_center_finder` -> `beam`?

After all, PEP8 says: "Modules should have short, all-lowercase names. Underscores can be used in the module name if it improves readability."

Comment (nvaytet) [reply docs/api-reference/index.md:32]: Or `beamcenter`? I feel `beam` is a little too vague?

Comment (SimonHeybrock) [reply docs/api-reference/index.md:32]: I suggest we keep it for now. We should probably sit down with Wojciech to review all the naming in the repo.

**Review by nvaytet — APPROVED** [review]

**Pull request merged**